packages/dds/merge-tree/src/test/client.spec.ts

```diff
@@ -207,6 +207,324 @@ describe("TestClient", () => {
+				});
+			}
+			for (let index = client.getLength() - 1; index >= 0; index--) {
+				const tile = client.searchForTile(index, tileLabel, true);
+
+				assert(tile, `Returned tile undefined @ ${index}.`);
+
+				assert.equal(
+					tile.pos,
+					index - (index % 3),
+					`Tile with label not at expected position: index ${index}, tile.pos ${tile.pos}`,
+				);
+			}
+		});
+
+		it("Should be able to find tile from client with text length 1", () => {
+			const tileLabel = "EOP";
+			client.insertMarkerLocal(0, ReferenceType.Tile, {
+				[reservedTileLabelsKey]: [tileLabel],
+				[reservedMarkerIdKey]: "some-id",
+			});
+
+			console.log(client.getText());
+
+			assert.equal(client.getLength(), 1, "length not expected");
+
+			const tile = client.searchForTile(client.getLength() - 1, tileLabel, false);
+
+			assert(tile, "Returned tile undefined.");
+
+			assert.equal(tile.pos, 0, "Tile with label not at expected position");
+
+			const tile1 = client.searchForTile(0, tileLabel, true);
+
+			assert(tile1, "Returned tile undefined.");
+
+			assert.equal(tile1.pos, 0, "Tile with label not at expected position");
+		});
+
+		it("Should be able to find only preceding but not non preceding tile with index out of bound", () => {
+			const tileLabel = "EOP";
+			client.insertMarkerLocal(0, ReferenceType.Tile, {
+				[reservedTileLabelsKey]: [tileLabel],
+				[reservedMarkerIdKey]: "some-id",
+			});
+
+			client.insertTextLocal(0, "abc");
+			console.log(client.getText());
+
+			assert.equal(client.getLength(), 4, "length not expected");
+
+			const tile = client.searchForTile(5, tileLabel, false);
+
+			assert.equal(typeof tile, "undefined", "Returned tile should be undefined.");
+
+			const tile1 = client.searchForTile(5, tileLabel);
+
+			assert.equal(typeof tile1, "undefined", "Returned tile should be undefined.");
+
+			const tile2 = client.searchForTile(-1, tileLabel);
+
+			assert.equal(typeof tile2, "undefined", "Returned tile should be undefined.");
+		});
+
+		it("Should return undefined when trying to find tile from text without the specified tile", () => {
+			const tileLabel = "EOP";
+			client.insertTextLocal(0, "abc");
+			console.log(client.getText());
+
+			assert.equal(client.getLength(), 3, "length not expected");
+
+			const tile = client.searchForTile(1, tileLabel);
+
+			assert.equal(typeof tile, "undefined", "Returned tile should be undefined.");
+
+			const tile1 = client.searchForTile(1, tileLabel, false);
+
+			assert.equal(typeof tile1, "undefined", "Returned tile should be undefined.");
+		});
+
+		it("Should return undefined when trying to find tile from null text", () => {
+			const tileLabel = "EOP";
+
+			const tile = client.searchForTile(1, tileLabel);
+
+			assert.equal(typeof tile, "undefined", "Returned tile should be undefined.");
+
+			const tile1 = client.searchForTile(1, tileLabel, false);
+
+			assert.equal(typeof tile1, "undefined", "Returned tile should be undefined.");
+		});
+	});
```

Comment:
Haven't read all of these test cases yet, but have you checked how well they verify coverage? Whenever adding new functionality it's interesting to consider what the coverage results look like if you only run the test suite for that functionality. I would hope that this suite covers all paths inside the function you added, and most branch paths inside `depthFirstNodeWalk`.

Reply:
All of the new search function is covered, and the only portion of depthFirstNodeWalk that is not covered is portion dealing with an undefined leafAction.